### ISSUE TITLE
Too many vertices in ImDrawList

### ISSUE THREAD

**Issue by fenix31415** [Fattorino/ImNodeFlow#17]
First, thank you for this awesome project!

The problem is that I want to deal with many vertices. This is why I get the [following assertion error](https://github.com/ocornut/imgui/blob/da29b776eed289db16a8527e5f16a0e1fa540251/imgui_draw.cpp#L2205).

As the comment suggests, it is a good idea to skip drawing nodes that are outside of the editor view.

I am not quite familiar with DearImgui, in particular with this project -- I just started using it. But I think that it is possible to skip calls [here](https://github.com/Fattorino/ImNodeFlow/blob/master/src/ImNodeFlow.cpp#L275): before `update`, it can check whether a node inside the render window and draw it only if it is in inside.

If this edit does not breaks imgui/ImNodeFlow architecture/invariants, I even can try to implement it myself and create a PR.

Thank you!

**Comment by Fattorino** [comment]
Creating a clip area and only rendering what's visible is the approach I prefer. For this what will need to happen is... first split the update function is a logic update and a render; then call the logic update each frame and the render only if necessary.

If you want to create a PR that follows this philosophy I'll be more than happy to review it and hopefully merge it.
Unfortunately, I won't be able to fix this issue before the weekend, even tho it's a very serious bug.

Let me know if you decide to work on it yourself so we don't end up working on the same thing.

**Comment by fenix31415** [comment]
Thank you for answer!

There is no rush, I will try to implement it.

Do you mean by "creating a clip area" something specific from Imgui or just a "virtual" rectangle that is being drawn?

**Comment by Fattorino** [comment]
just a custom imaginary rectangle, nothing specific with ImGui

**Comment by fenix31415** [comment]
I did something and it works. 
[Here](https://github.com/fenix31415/ImNodeFlow/tree/clip_area) you can look at my branch (2 commits).

Unfortunately, I can't say for sure which code from `BaseNode::Update` is related to processing and which is related to drawing. I can assume that there is something connected with passing values in links. But I do not even know how to test it.

Anyway, this code works for me -- I skip all update entirely. As I said before, I use only nodes and links, `behavior` is always empty function for me.

I do not think it deserves a PR -- I couldn't split to the logic update and a render update. Anyway, you could look at the code and I hope it helps you. At least, with detecting a "clip area".

Also, I have no idea what `win_pos` is (added to the grid lines). It is always zero for me. Likely, you need to take this variable into account too (I just cannot test how it affects).

**Comment by Fattorino** [comment]
I will look into it (and I already left a comment on the PR), and I'll finish working on it this weekend.
As per the `win_pos`, yes it's always 0 because it's dead code now that I use a separate context for rendering everything. I'll remove it now :)

**Comment by Fattorino** [comment]
One question, how exactly did you run into this error? Just by adding a lot of nodes?

**Comment by fenix31415** [comment]
Yes. ~500 nodes and ~500 links. 
Also, with my fix, I can zoom to see all graph and again get an error.

**Comment by fenix31415** [comment]
In case someone runs into same problem.
Even when rendering only visible nodes, I get the error, as I mentioned before. I still want to zoom out to see a whole graph.

I used ImGui debug window to see all used indices. It turned out that some elements takes too many of them. It is easier to render rectangles (two triangles) instead of circles (too many triangles).
I found the following ones:

- Rounded corners. I set `NodeStyle::radius` zero.
- Pins. I used a custom renderer which just uses `AddNgonFilled` for 4 segments.
- Connections. It is a bezier, which contains many triangles. Unfortunately there is hardcoded (look at `smart_bezier(..)`) `delta` variable. I set it to 0 in the library directly. Connections became straight lines.
- Text. Every symbol costs 2 triangles. I disabled rendering text with small zoom.

Those optimizations (+ rendering only visible nodes, of course) made possible to see almost entire graph.

**Comment by Fattorino** [comment]
Yes I'm aware of the need to simplify rendering at small scales, and I'm working on a way to simplify rendering proportionally to the zoom level in order to keep as much details as possible. The idea is that the user can select the detail scaling level based on his needs and the number of nodes

**Comment by Fattorino** [comment]
with commit b9ded55 I implemented what your comment says, just removing a lot of stuff when zoomed out, in the future I plan to create something a little more flexible but now I have very little time.

Zoomed in:
![image](https://github.com/Fattorino/ImNodeFlow/assets/90210751/c6ead8ef-a954-4bb2-80d3-a5c53f5caf1b)

Zoomed out (< 0.7):
![image](https://github.com/Fattorino/ImNodeFlow/assets/90210751/9ad5225d-08ff-4b05-ad1a-1dce5a3c141a)

It's a very drastic cut that's why in the future I want to implement something more gradual

(I'm still working on the clip rectangle because it's not as easy as I thought to calculate what nodes must be rendered)

**Comment by fenix31415** [comment]
Cool! It would be cool to have opportunity to tweak the zoom border.

Speaking on clip rectangle, you can look at my code. I even draw the rectangle to test that all right.

**Comment by Fattorino** [comment]
> tweak the zoom border

What do you mean by that?

<BR>Also, could you send me the link to the clip-rect code again, please?

**Comment by fenix31415** [comment]
You said the current zoom border is 0.7 now. I think it shouldn't be a hardcoded constant, but lay somewhere at style.

Clip rectangle code:https://github.com/fenix31415/ImNodeFlow/blob/clip_area/src/ImNodeFlow.cpp#L308

**Comment by Fattorino** [comment]
with the reduced rendering, without the clip-rect, doest it crash?
(Also, I'm still working on the clip-rect. But updating the logic to allow data flow without rendering is not as easy as I thought so I'm doing some refactoring)

**Comment by fenix31415** [comment]
Yes it still crash. But it definitely lets render mode nodes (300-400). Before this commit I could render only ~100. In total, I have 550 of them for now.

Take your time about clip rectangle. I just use my code (for clip rectangle) for now + disallowed full zoom (where I can see whole graph).

**Comment by Fattorino** [comment]
I created a "dev" branch with the clip-rect code. I have some problems determining correctly these two variables:
https://github.com/Fattorino/ImNodeFlow/blob/a214631530159cbeb42807ce91103e28c1b6cf83/src/ImNodeFlow.cpp#L289
https://github.com/Fattorino/ImNodeFlow/blob/a214631530159cbeb42807ce91103e28c1b6cf83/src/ImNodeFlow.cpp#L290

**Comment by Fattorino** [comment]
I was also thinking about adding this `#define ImDrawIdx unsigned int`. Having 32-bit indices just makes sense

**Comment by Fattorino** [comment]
@fenix31415 can you check if it crashes using larger indices?

**Comment by fenix31415** [comment]
Where should I define it?

**Comment by Fattorino** [comment]
Im not sure what's the best way, you can read the comment near the assert

**Comment by Fattorino** [comment]
Seeing as now it's fairly common to build with the extended vertices index, I feel this is not really an issue anymore. Although clipping would still be a very nice feature
